### PR TITLE
[Speculative] DFlash tree drafting: host-side tree-construction core (#29524)

### PR DESCRIPTION
**Pull request by ajinkyajawale14499**
## Motivation

This is the first step toward **JetSpec speculative decoding** (#29524). SGLang's **DFlash** drafts a **linear chain** — one candidate token per draft depth, scored independently of which branch it sits on. JetSpec (Hao AI Lab, [arXiv 2606.18394](https://arxiv.org/abs/2606.18394)) keeps DFlash's single draft-head forward pass but proposes a **token tree** and verifies the whole tree in one target pass under a tree-causal mask.

The core problem with the chain: **it cannot use a larger draft budget productively.** Once you draft more tokens linearly, acceptance saturates, because a single mispredicted token truncates the whole proposal. A tree spends the same budget on *branches*, so more of it survives verification. Concretely, from the paper (Qwen3-8B, temperature 0):

| Drafting | Accept length τ @ budget 256 (MATH-500) | Speedup |
|---|---|---|
| DFlash chain | saturates ~budget 32 (τ≈6.9), no gain beyond | — |
| EAGLE-3 tree (depth 8, budget 64) | τ≈4.13 | 2.36× |
| **JetSpec tree** | **τ≈10.76** | **9.64×** |

(JetSpec also reports ~1000 TPS on a single B200 and 9.45× on MoE Qwen3-30B-A3B.) The win is structural: same one-pass draft head, but branch scores are conditioned on ancestors and verified as a tree.

This is a large, GPU-bound feature (verify-loop integration, CUDA-graph capture, multiple attention backends, kernels, perf). **This PR lands only the GPU-independent host-side core + plumbing, fully CPU-unit-tested**, so the verify-loop integration can be built on a tested foundation. Tree mode is guarded off at startup; the chain path is byte-for-byte unchanged.

I want to be transparent up front: **I don't currently have GPU access**, so I deliberately stopped at the GPU boundary rather than ship untested kernel/worker code (see Accuracy Tests / Speed Tests). I'd welcome collaboration on the verify-loop wiring and validation — @vincentzed.

## Background — how the tree is built (1 minute)

Per request, the draft head's one forward pass gives per-depth logits. From these we take **top-W tokens + log-probs per depth** (`(depth, W)`), pulled to host. The tree is then built **entirely on CPU**:

- **`opt_prefix`** (paper Algorithm 1): a best-first max-heap keyed by accumulated branch log-prob `Σ log q(y_u | x, h, ancestors)`. Each pop emits one node and pushes its next sibling (same parent, next rank) and first child (next depth, rank 0) — enumerating exactly the top-B highest-probability root→node prefixes without expanding the exponential space.
- **`depth_first`** (default): lays down the greedy top-1 chain (the "spine") first, then adds side branches by accumulated log-prob. Because the spine is always present, **tree acceptance ≥ chain acceptance** by construction — a safe no-regression floor.

The tree is then encoded for verification: a **tree-causal "allow" mask** where each node attends only to the prefix + its ancestors (not siblings/cousins), and SGLang's existing **`retrieve_next_token`/`retrieve_next_sibling`** left-child/right-sibling pointers that the tree-general verify kernels already walk. Acceptance is greedy: a child is accepted iff its token equals the target's argmax at its parent; the longest fully-accepted root→leaf path wins.

This is a faithful port of the reference vLLM integration ([`Jet-Flow/vllm-jetspec`](https://github.com/Jet-Flow/vllm-jetspec)), adapted to SGLang's DFlash/EAGLE conventions.

## Modifications

**New `python/sglang/srt/speculative/dflash_tree.py`** — pure-Python/torch, no sglang-internal imports, no GPU:
- `build_tree_opt_prefix`, `build_tree_depth_first`, and a `build_tree` dispatcher (caps fan-out to `tree_width`).
- `tree_to_verify_buffers` — tree → SGLang verify buffers (`retrieve_next_token`/`retrieve_next_sibling` with the `-1` sentinel, `positions`, padded to budget).
- `build_tree_custom_mask` / `build_ancestor_mask` — tree-causal mask in the existing flattened `Σkv·N + N²·bs` layout.
- `compute_tree_budget`, `DFlashDraftTree` (`msgspec.Struct`), and `verify_tree_greedy_cpu` (CPU reference of greedy tree acceptance — test oracle + eager fallback).

**Plumbing:**
- `server_args.py` — `--speculative-dflash-tree-{width,budget,construction}` (documented as guarded/inert until the verify-loop integration lands).
- `arg_groups/speculative_hook.py` — `_handle_dflash_tree` rejects `--speculative-dflash-tree-width > 1` at startup with a clear message, so the flag can't trigger an unimplemented path. `width <= 1` (chain) is unaffected.
- `dflash_info.py` — `DFlashVerifyInput.from_tree` (populates the existing `topk`/`custom_mask` fields) and `build_dflash_tree_verify_input` (draft top-W output → tree-mode verify input): the seam the worker integration will call.
- `test/registered/spec/dflash/test_dflash_tree.py` — CPU unit + integration suite (`register_cpu_ci`, `base-a-test-cpu`).

**Key design decisions:**
1. **Reuse the tree-general verify kernels** (`verify_tree_greedy` / `tree_speculative_sampling_target_only`), **not** `build_tree_kernel_efficient` — a best-first tree is general and doesn't fit EAGLE's per-step beam `parent_list` format, whereas the verify kernels consume our hand-built `retrieve_*` + `custom_mask` directly. **No new kernels.**
2. **`tree_width == 1` is byte-identical to the current chain buffers** (`retrieve_next_token = t→t+1`, last `-1`; `retrieve_next_sibling` all `-1`) — asserted in a test against the actual `_get_or_create_chain_verify_buffers`. Existing DFlash users are unaffected.
3. **The DFlash worker decode loop is intentionally untouched** (zero risk to the chain path); tree mode is wired through the new bridge in a follow-up.

**Deferred (follow-up, GPU-bound):** wiring the tree path into `dflash_worker_v2`'s verify loop — CUDA-graph capture for a *variable-size* tree, and the **non-contiguous accepted-path KV commit** (the chain path assumes a contiguous accepted prefix); plus `trtllm_mha`/`flashinfer` tree-verify, MoE/hybrid recurrent-tree commit, SM90/SM100, rejection-sampling tree, and perf.

## Accuracy Tests

This PR does **not change model outputs**: tree mode is guarded off at startup, and the chain path is byte-identical (asserted). So there is no accuracy delta to report here.

The host-side core is validated by a CPU-only suite (no GPU, no server):
- **Keystone integration test** — a faithful CPU simulation of the CUDA `verify_tree_greedy` traversal (follow `retrieve_next_token` → `retrieve_next_sibling`) is asserted **equal** to an independent parent-based acceptance oracle over **200 random trees × both builders**. This validates that the buffer encoding the GPU kernel will consume matches the tree the host built — the central correctness argument that substitutes for a GPU run.
- **W=1 byte-equivalence** to the legacy chain buffers (via the real `_get_or_create_chain_verify_buffers`).
- Builder invariants (parent < child, depth consistency, budget cap, opt_prefix best-first ordering, depth_first spine guarantee), tree-causal mask semantics + size formula, full edge-case coverage (empty depth, budget 0/1, budget > full-tree, single depth, ties, padding, batch), and the startup-guard + bridge.

The **end-to-end lossless-greedy accuracy test** — tree-spec output token-identical to non-spec decode on `triton`/`fa3` with `JetSpec/jetspec-qwen3-8b` — is the merge gate for the follow-up verify-loop wiring. It requires a GPU (which I don't currently have); the code is structured (bridge + oracle) so it's straightforward to validate once the loop is wired, and I'd appreciate a hand running it.

## Speed Tests and Profiling

No speed impact in this PR — tree mode is guarded off and the chain path is unchanged. Throughput / acceptance-length benchmarking (tree vs chain across budgets; SM90/SM100) is part of the deferred follow-up and requires a GPU; not included here.

## Checklist

- [x] Format code with pre-commit (black + isort + ruff lint; verified clean).
- [x] Add unit tests (CPU unit + integration suite under `test/registered/spec/dflash/`).
- [ ] Update documentation — N/A for this PR (no user-facing feature yet; flags documented in `--help` and marked inert/guarded).
- [ ] Provide accuracy and speed benchmark results — deferred: tree mode isn't runnable yet (guarded) and I have no GPU; this is the follow-up's merge gate (see above).
- [x] Follow the SGLang code style guidance.

## References

- Issue: #29524 · Paper: [arXiv 2606.18394](https://arxiv.org/abs/2606.18394) · Blog: https://haoailab.com/blogs/parallel-tree-decoding
- Reference integration: [`Jet-Flow/vllm-jetspec`](https://github.com/Jet-Flow/vllm-jetspec) · Draft heads: https://huggingface.co/JetSpec

Resolves part of #29524.

---
### CI States

Latest PR Test (Base): :x: [Run #28689534811](https://github.com/sgl-project/sglang/actions/runs/28689534811)
Latest PR Test (Extra): :x: [Run #28689534729](https://github.com/sgl-project/sglang/actions/runs/28689534729)

### CODE REVIEW

**Comment by ajinkyajawale14499** [comment]
Thanks for the review — all three items addressed:

- **width==0 guard** (`build_tree_depth_first`) and **`target_predict` tensor→list** (`verify_tree_greedy_cpu`): applied via the suggestions.
- **Remaining item — tensor inputs to the builders**: `build_tree` now normalizes tensor / list-of-tensor inputs to host lists up front (`_rows_to_lists`), so the per-node Python loops avoid torch scalar-dispatch overhead. This covers both builders at the single production entry point (the bridge / worker call `build_tree`, not the inner builders directly).
- Added tests: a zero-width-rows regression, plus tensor-input coverage for `build_tree` and the accept oracle.

Still a draft — the end-to-end lossless-greedy validation is the GPU follow-up (no GPU on my end yet).

**Comment by ronhuafeng** [comment]
I took a pass over draft head `6694c497cf0ab3dc669ef1c36197fa48314e9b64`.

Two things look actionable before the GPU follow-up:

- The current GitHub lint failure is a real changed-file failure. In the dev container, changed-file pre-commit reproduced it: ruff rewrites the two `DFlashVerifyInput` return annotations in `python/sglang/srt/speculative/dflash_info.py`, and codespell flags the `anc` local name in `python/sglang/srt/speculative/dflash_tree.py` lines 263/265. Renaming it to something like `ancestor_mask` fixes the codespell part.
- The focused CPU suite currently has one real failure: `pytest test/registered/spec/dflash/test_dflash_tree.py -q` gives `47 passed, 1 failed` at `TestTreeMask.test_context_all_allowed`.

For the mask failure, the generated mask for `DFlashDraftTree([0, 1], [-1, 0], [0, 1], 2)` with `kv_lens=[3]` is row-major per query row:

```python
mask.tolist() == [True, True, True, True, False, True, True, True, True, True]
mask.view(2, 5).tolist() == [
    [True, True, True, True, False],
    [True, True, True, True, True],
]
```

That matches the Triton custom-mask indexing path (`row * cur_seq_kv_len + col`). So I think the implementation layout is consistent with the backend reader, while the test assertion is slicing as if all prefix cells were contiguous at the front. Locally, changing that assertion to check the prefix columns per row, e.g. `mask.view(2, 3 + 2)[:, :3].all()`, plus the lint fixes above, gives:

- `pytest test/registered/spec/dflash/test_dflash_tree.py -q` -> `48 passed`
- changed-file `pre-commit run --files ...` -> passed

I did not run the GPU verify-loop wiring; this is only CPU/static CI triage for the draft PR.

**Comment by ajinkyajawale14499** [comment]
Pushed `b828b32` addressing all of @ronhuafeng's triage — thanks, this was spot-on:

- *codespell*: `anc` → `ancestor_mask`.
- *ruff*: dropped the redundant quotes on the two `DFlashVerifyInput` return annotations (lazy under `from __future__ import annotations`).
- *`test_context_all_allowed`*: your diagnosis was exactly right — the mask is row-major per query row (`[kv context | N block]`, matching the Triton `row*kv_len+col` reader), so the *implementation* layout is correct and the *assertion* was slicing a flat prefix. Fixed it to check the context columns per row (`mask.view(n, kv+n)[:, :kv]`). Local `pre-commit run --files …` is now green; the test fix matches what you verified (48 passed).

For context (in case anyone's landing here fresh): this PR is the **GPU-independent host-side core** for JetSpec tree drafting (#29524) — tree builders (`opt_prefix` / `depth_first`), the `retrieve_*` + tree-causal-mask buffer encoding, a CPU accept oracle, and the `--speculative-dflash-tree-*` flags. All CPU-unit-tested; tree mode is *guarded off at startup* and the chain path is byte-identical, so it's zero-risk to existing DFlash. The live `dflash_worker_v2` verify-loop wiring + e2e lossless validation are the GPU follow-up, being coordinated with @vincentzed (per his note on #29524 — collaborating via Slack).

File: python/sglang/srt/speculative/dflash_tree.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+    budget,
+    construction="depth_first",
+):
+    """Dispatch to a builder using at most ``tree_width`` candidates per depth.
+
+    Rows may carry more top-k columns than wanted, so cap the per-depth fan-out to
+    ``tree_width`` here. ``tree_width == 1`` degenerates to a chain in either builder.
+    """
+    topk_tokens = [row[:tree_width] for row in topk_tokens]
+    topk_logprobs = [row[:tree_width] for row in topk_logprobs]
```
[the remark this answers is not in the document]
Comment (ajinkyajawale14499) [reply python/sglang/srt/speculative/dflash_tree.py:177]: build_tree normalizes tensor inputs to host lists before the slice, via _rows_to_lists:

topk_tokens = _rows_to_lists(topk_tokens)
topk_logprobs = _rows_to_lists(topk_logprobs)
topk_tokens = [row[:tree_width] for row in topk_tokens]
topk_logprobs = [row[:tree_width] for row in topk_logprobs]

def _rows_to_lists(rows):
    if torch.is_tensor(rows):
        return rows.tolist()
    return [row.tolist() if torch.is_tensor(row) else list(row) for row in rows]

This covers all three cases (tensor / list-of-tensors / list-of-lists) at build_tree, which is the single production entry point; the bridge (build_dflash_tree_verify_input) and the worker call build_tree, so both builders always receive nested Python lists and never do per-node scalar dispatch on tensors. The inner build_tree_* builders are only called directly from CPU unit tests (with lists). This landed in an earlier revision than the one your comment is anchored to.
